File: Patterns/Command Pattern/Cpp/Command/Stereo/StereoOffCommand.cpp
#include "../ICommand.cpp"
#include "./Stereo.cpp"

class StereoOffCommand : public ICommand {
private:
    Stereo* stereo;
    int preview;

public:
    StereoOffCommand(Stereo* stereo) : stereo{stereo}, preview{stereo->OFF} {}

    void execute() override {
        preview = stereo->getState();
        stereo->off();
    }

    void undo() override {
        if (preview == stereo->CD) {
            stereo->on();
            stereo->setCd();
            stereo->setVolumn(11);
        }
        else if (preview == stereo->DVD) {
            stereo->on();
            stereo->setDvd();
            stereo->setVolumn(11);
        }
        else if (preview == stereo->RADIO) {
            stereo->on();
            stereo->setRadio();
            stereo->setVolumn(11);
        }
    }

    string getCommandName() override {
        return "StereoOffCommand";
    }
};

File: Patterns/Command Pattern/CSharp/Command/Stereo/StereoOffCommand.cs
namespace Command.Stereo
{
    public class StereoOffCommand : ICommand
    {
        private Stereo stereo;
        private int preview;

        public StereoOffCommand(Stereo stereo)
        {
            preview = Stereo.OFF;
            this.stereo = stereo;
        }

        public void Execute()
        {
            preview = stereo.State;
            stereo.Off();
        }

        public void Undo()
        {
            switch (preview)
            {
                case Stereo.CD:
                    stereo.On();
                    stereo.SetCd();
                    stereo.SetVolumn(11);
                    break;
                case Stereo.DVD:
                    stereo.On();
                    stereo.SetDvd();
                    stereo.SetVolumn(11);
                    break;
                case Stereo.RADIO:
                    stereo.On();
                    stereo.SetRadio();
                    stereo.SetVolumn(11);
                    break;
            }
        }
    }
}